File: modules/androidkit/src/SkottieAnimation.cpp
/*
 * Copyright 2021 Google Inc.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

#include <jni.h>

#include "modules/androidkit/src/Utils.h"
#include "modules/skottie/include/Skottie.h"

using namespace skottie;

namespace {

static jlong Animation_Create(JNIEnv* env, jobject, jstring jjson) {
    const androidkit::utils::CString cstr(env, jjson);

    // TODO: more builder opts
    auto animation = Animation::Builder().make(cstr, strlen(cstr));

    return reinterpret_cast<jlong>(animation.release());
}

static void Animation_Release(JNIEnv, jobject, jlong native_animation) {
    SkSafeUnref(reinterpret_cast<Animation*>(native_animation));
}

static jdouble Animation_GetDuration(JNIEnv, jobject, jlong native_animation) {
    const auto* animation = reinterpret_cast<const Animation*>(native_animation);
    return animation ? animation->duration() : 0;
}

static jdouble Animation_GetFrameCnt(JNIEnv, jobject, jlong native_animation) {
    const auto* animation = reinterpret_cast<const Animation*>(native_animation);
    return animation ? animation->outPoint() : 0;
}

static jfloat Animation_GetWidth(JNIEnv, jobject, jlong native_animation) {
    const auto* animation = reinterpret_cast<const Animation*>(native_animation);
    return animation ? animation->size().width() : 0;
}

static jfloat Animation_GetHeight(JNIEnv, jobject, jlong native_animation) {
    const auto* animation = reinterpret_cast<const Animation*>(native_animation);
    return animation ? animation->size().height() : 0;
}

static void Animation_SeekFrame(JNIEnv, jobject, jlong native_animation, jdouble frame) {
    if (auto* animation = reinterpret_cast<Animation*>(native_animation)) {
        animation->seekFrame(frame);
    }
}

static void Animation_SeekTime(JNIEnv, jobject, jlong native_animation, jdouble t) {
    if (auto* animation = reinterpret_cast<Animation*>(native_animation)) {
        animation->seekFrameTime(t);
    }
}

static void Animation_Render(JNIEnv, jobject, jlong native_animation, jlong native_canvas) {
    const auto* animation = reinterpret_cast<const Animation*>(native_animation);
    auto* canvas = reinterpret_cast<SkCanvas*>(native_canvas);
    if (animation && canvas) {
        animation->render(canvas);
    }
}

} // namespace

int register_androidkit_SkottieAnimation(JNIEnv* env) {
    static const JNINativeMethod methods[] = {
        {"nCreate"       , "(Ljava/lang/String;)J", reinterpret_cast<void*>(Animation_Create)     },
        {"nRelease"      , "(J)V"                 , reinterpret_cast<void*>(Animation_Release)    },

        {"nGetDuration"  , "(J)D"                 , reinterpret_cast<void*>(Animation_GetDuration)},
        {"nGetFrameCount", "(J)D"                 , reinterpret_cast<void*>(Animation_GetFrameCnt)},
        {"nGetWidth"     , "(J)F"                 , reinterpret_cast<void*>(Animation_GetWidth)   },
        {"nGetHeight"    , "(J)F"                 , reinterpret_cast<void*>(Animation_GetHeight)  },

        {"nSeekFrame"    , "(JD)V"                , reinterpret_cast<void*>(Animation_SeekFrame)  },
        {"nSeekTime"     , "(JD)V"                , reinterpret_cast<void*>(Animation_SeekTime)   },
        {"nRender"       , "(JJ)V"                , reinterpret_cast<void*>(Animation_Render)     },
    };

    const auto clazz = env->FindClass("org/skia/androidkit/SkottieAnimation");
    return clazz
        ? env->RegisterNatives(clazz, methods, SK_ARRAY_COUNT(methods))
        : JNI_ERR;
}

File: platform_tools/android/apps/AndroidKit/src/main/java/org/skia/androidkit/SkottieAnimation.java
/*
 * Copyright 2021 Google Inc.
 *
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

package org.skia.androidkit;

import org.skia.androidkit.Canvas;

public class SkottieAnimation {
    private long mNativeInstance;

    /**
      * Create an animation from the provided JSON string.
      */
    public SkottieAnimation(String animation_json) {
        mNativeInstance = nCreate(animation_json);
    }

    /**
      * Returns the animation duration in seconds.
      */
    public double getDuration() {
        return nGetDuration(mNativeInstance);
    }

    /**
     * Returns the animation frame count.  This is normally an integral value,
     * but both the JSON encoding and Skottie's frame-based APIs support fractional frames.
     */
    public double getFrameCount() {
        return nGetFrameCount(mNativeInstance);
    }

    /**
      * Returns the intrinsic animation width.
      */
    public float getWidth() {
        return nGetWidth(mNativeInstance);
    }

    /**
      * Returns the intrinsic animation height.
      */
    public float getHeight() {
        return nGetHeight(mNativeInstance);
    }

    /**
      * Update the animation state to match |t|, specifed in seconds.
      * The input is clamped to [0..duration).
      */
    public void seekTime(double t) {
        nSeekTime(mNativeInstance, t);
    }

    /**
     * Update the animation state to match |f|, specified as a frame index
     * in the [0..frameCount) domain.
     *
     * Fractional values are allowed and meaningful - e.g.
     *
     *   0.0 -> first frame
     *   1.0 -> second frame
     *   0.5 -> halfway between first and second frame
     */
    public void seekFrame(double f) {
        nSeekFrame(mNativeInstance, f);
    }

    /**
      * Draw the current frame to the Canvas.
      */
    public void render(Canvas canvas) {
        nRender(mNativeInstance, canvas.getNativeInstance());
    }

    /**
     * Releases any resources associated with this Animation.
     */
    public void release() {
        nRelease(mNativeInstance);
        mNativeInstance = 0;
    }

    @Override
    protected void finalize() throws Throwable {
        release();
    }

    private static native long nCreate(String json);
    private static native void nRelease(long nativeInstance);

    private static native double nGetDuration(long nativeInstance);
    private static native double nGetFrameCount(long nativeInstance);
    private static native float  nGetWidth(long nativeInstance);
    private static native float  nGetHeight(long nativeInstance);

    private static native void nSeekTime(long nativeInstance, double t);
    private static native void nSeekFrame(long nativeInstance, double frame);
    private static native void nRender(long nativeInstance, long nativeCanvas);
}
